Budget component with many budgets

agustibr opened the pull request:
#### :tophat: What? Why?

Refactor the Budget component to create a `Decidim::Budgets::Budget` resource, and make `Decidim::Budgets::Project`s depend on it. Each component can have many “Budgets”, thus allowing for multiple budget resources in a single component.

#### :pushpin: Related Issues
- Related to #5993
- Depends on: #6274 
- Fixes #5711

#### :clipboard: Subtasks
- [ ] Add `CHANGELOG` upgrade notes, if required
- [x] If there's a new public field, add it to GraphQL API
- [x] Add/modify seeds


- [x] 1. add budget model
  - [x] add migration and model
  - [x] admin routes, controllers, commands and forms
  - [x] create front end routes and controllers
  - [x] apply home design
  - [x] add budgets workflow
  - [x] one budget? redirect index to that budget
  - [x] Add tests
  - [x] add graphql
  - [x] add searchable
- [x] 2. modify projects to use the budgets model
  - [x] migrations and models
  - [x] admin routes, controllers, commands and forms
  - [x] folders and files
  - [x] import proposals to projects
  - [x] front-end routes, controllers
  - [x] update graphql, search
  - [x] fix tests
- [x] 3.modify orders to use the budgets model
  - [x] migrations and models
  - [x] front-end routes, controllers and commands
  - [x] update data portability
- [x] 4. migrate existing data
- [x] 5. complex votings
  - [x] workflows done to the new scenario
  - [x] Add workflow documentation



### :camera: Screenshots (optional)

- Budgets index (Vote in all workflow):
  <img width="707" alt="Screenshot 2020-07-15 at 17 34 32" src="https://user-images.githubusercontent.com/210216/87565112-8163a380-c6c1-11ea-9670-39d4a6d301bd.png">

- Budgets index with highlighted resource:
  <img width="949" alt="Screenshot 2020-07-15 at 17 50 59" src="https://user-images.githubusercontent.com/210216/87566878-e0c2b300-c6c3-11ea-86cd-d02ad4a58c5c.png">

- Searchable Budget resource:  
  <img width="801" alt="Screenshot 2020-08-27 at 15 58 22" src="https://user-images.githubusercontent.com/210216/91540975-e72b8800-e91b-11ea-8f2f-250da8810bb0.png">

### Project admin options | user's current order state

1.  voting enabled | order not checked out:
  <img width="865" alt="Screenshot 2020-07-15 at 17 20 56" src="https://user-images.githubusercontent.com/210216/87563958-1bc2e780-c6c0-11ea-886b-08aaa49cac4f.png">
2. voting enabled + show votes | order not checked out:
  <img width="856" alt="Screenshot 2020-07-15 at 17 23 03" src="https://user-images.githubusercontent.com/210216/87564137-53ca2a80-c6c0-11ea-8695-5f6f417dbd4b.png">
3. voting enabled + show votes | checked out order:
  <img width="852" alt="Screenshot 2020-07-15 at 17 27 37" src="https://user-images.githubusercontent.com/210216/87564433-ae638680-c6c0-11ea-928e-417bf331790d.png">
4. show votes | checked out order:
  <img width="855" alt="Screenshot 2020-07-15 at 17 30 09" src="https://user-images.githubusercontent.com/210216/87564594-de128e80-c6c0-11ea-9023-b549b99f1766.png">

### admin panel UI

1. New/Edit budget component has workflows to choose from:  
  <img width="776" alt="Screenshot 2020-07-29 at 16 12 20" src="https://user-images.githubusercontent.com/210216/88813593-67de5380-d1b9-11ea-9189-cc9bb1125f2a.png">
2. Manage Budgets, new budget button and list budget resources:  
  <img width="784" alt="Screenshot 2020-07-29 at 16 15 13" src="https://user-images.githubusercontent.com/210216/88813797-a70ca480-d1b9-11ea-8f3a-0452344073dc.png">
3. Manage Projects within a budget resource:  
  <img width="783" alt="Screenshot 2020-07-29 at 16 15 34" src="https://user-images.githubusercontent.com/210216/88813889-c1468280-d1b9-11ea-89b2-04a9349287eb.png">

leio10 commented on decidim-budgets/app/types/decidim/budgets/budget_type.rb:
```diff
@@ -0,0 +1,23 @@
+      name "Budget"
+      description "A budget"
+
+      field :id, !types.ID, "The internal ID of this election"
```
"of this budget" :)

agustibr replied:
Sure! 😀

leio10 commented on decidim-budgets/app/types/decidim/budgets/project_type.rb:
```diff
@@ -16,7 +16,7 @@ module Budgets
       field :title, Decidim::Core::TranslatedFieldType, "The title for this project"
       field :description, Decidim::Core::TranslatedFieldType, "The description for this project"
-      field :budget, types.Int, "The budget for this project"
+      field :budget_amount, types.Int, "The budget for this project"
```
I think that the description "The budget for this project" could be used for a field with a reference to the budget object and for this field. Maybe "The budget amount for this project" would avoid this?

agustibr replied:
Yes, agree 👍

leio10 commented on decidim-budgets/db/migrate/20200714103519_move_budgets_to_own_model.rb:
```diff
@@ -0,0 +1,107 @@
+    self.table_name = :decidim_budgets_orders
+  end
+
+  def up
```
Very nice work here! 🔝 👏

leio10 commented on decidim-budgets/db/migrate/20200714103519_move_budgets_to_own_model.rb:
```diff
@@ -0,0 +1,107 @@
+  end
+
+  def create_budget_resource_from(component)
+    component_total_budget = (component["settings"]["global"]["total_budget"] if component["settings"]["global"].try(:key?, "total_budget"))
```
maybe you can use the `dig` method here: `component.dig("settings", "global", "total_budget")`

leio10 commented on decidim-budgets/db/migrate/20200714103519_move_budgets_to_own_model.rb:
```diff
@@ -0,0 +1,107 @@
+    budget_components.each do |component|
+      resource = create_budget_resource_from(component)
+
+      if resource
```
there is any possibility of getting an empty `resource` here?

agustibr replied:
I don't think so, if a resource fails on `create!` an exception will be thrown. I'll remove the `if resource`

tramuntanal commented:
Hi @agustibr I see you restored the `feat/component_setting_scope` branch from #5941 to create #6309 which merges into this PR. So this PR is in WIP no? I'm adding this label then.
Feel free to remove it and ping core when ready

agustibr commented:
Hi @tramuntanal! this PR can be merged into `develop` independently from #6309. 
Is the other way, #6309 starts from this code base, as it includes scopes for components including the budgets (from this PR).

This PR is currently waiting reviews and suggestions, I'm removing the WIP label. 😁

tramuntanal commented:
:+1:  @agustibr , let us know when it is approved by @decidim/product

andreslucena commented:
Hi, I've some doubts/questions/comments: 

#### 1. Admin explanation 

I was reviewing this on staging and I think that we need to explain much better how it's configured, for instance: 

![imatge](https://user-images.githubusercontent.com/717367/88641094-03d76480-d0bf-11ea-82b0-7a6dad22e531.png)

(It's just a quick example, but I'm pretty sure you could come up with much better explaining). 

#### 2. Random workflow

I don't know/understand how to use this. Does it make sense? What's the use case? 

#### 3. "Complex" voting / PAM2020 

Just to confirm it, the workflow "PAM2020" will be on the Decidim Barcelona repository? 

#### 4. Legacy 

What will happen with all the legacy Budgets? Will they be migrated to the new model and there'll be a button?

agustibr commented:
Hi @andreslucena, thanks for reviewing!

Sure, I can add some explanations to the admin.

Budget workflows lets admins pick, at the component level, how a user can participate in it.

The `one`, `all` and `random` are included in `decidim` as examples, also to ensure that the `Budgets::Workflows::Base` class provides multiple possibilities for customization.
  - `one`: This Workflow allows users to vote in any budget, but only in one.
  - `all`: This Workflow allows users to vote in all budgets.
  - `random`: This Workflow allows users to vote only in one budget, selected randomly.

Workflows can be added or removed in the application [initializer file](https://github.com/codegram/decidim-staging/blob/master/config/initializers/decidim.rb#L134-L135). 

Yes the `PAM2020` will be added to Barcelona's repo, [this is the PAM2020 workflow](https://github.com/codegram/decidim-staging/blob/master/lib/decidim_staging/budgets_workflow_pam2020.rb) used in staging, that uses the budget scope (feature _to be_ added in #6309, present in staging) combined with the participant verification scope, to highlight the budget in the user scope and allow also to vote in another.

This PR includes compatibility for legacy installations:
- data migration (`up` and `down`) so existing data will be moved to the new `budget` model.
- when there's only one budget in the component, it behaves as previously: the frontend budget list is skipped and the participant is redirected to the project list.
- `one` is the budgets workflow selected by default.

I hope I've cleared up your doubts. 😁

andreslucena commented:
>    one: This Workflow allows users to vote in any budget, but only in one.
>    all: This Workflow allows users to vote in all budgets.

Can you add these explanations on the admin panel UI please :pray:?

>    random: This Workflow allows users to vote only in one budget, selected randomly.

Can we remove the `random` workflow? I know participatory budgeting processes that might want `one` and `all`, but `random` no... Maybe if it's only an example it could be on decidim-budgets documentation as a theoretical example on how to work with workflows, but I'd prefer to not having it as an option on the default installation.   


> This PR includes compatibility for legacy installations:
(...)

Perfect!! That's just what I wanted to hear :rocket:

agustibr commented:
@andreslucena the PR is updated with your suggestions:

- Added workflow explanations on the admin panel UI  
  <img width="673" alt="Screenshot 2020-07-29 at 16 11 53" src="https://user-images.githubusercontent.com/210216/88812285-d6221680-d1b7-11ea-902a-5f48dc7a78e3.png">

- Removed the `:random` workflow from `decidim-budgets`. I kept the workflow (now in `decidim-generators`) as an example of a custom workflow referenced in the documentation, and is added to the `development_app` and `test_app`.

Hope it's fine now, so I can ask _core_ to review it 😄

ahukkanen commented:
I've tested this and overall it looks pretty good! Here is some initial feedback.

### Bugs

- Go to component permissions -> You will see a missing translation (I don't know if this is a bug in the budgets itself or in these changes)
- You will get an error ```undefined method \`budget' for #<Decidim::Budgets::Budget:0x00001a2b345678c9>```` with the following conditions:
  * Workflow: Vote in all
  * Add two budgets with some projects
  * Finish the voting in both budgets
  * Go back to budgets front page
  => Fix it by changing `budget.budget` to `budget.total_budget` in `decidim-budgets/app/cells/decidim/budgets/budgets_list/voted.erb`

### Feature requests

- I think it should be possible to programmatically "register" new budget workflows. Currently I need to override the whole method (`Decidim::Budgets.workflows`)
- If the workflow tells voting is not possible anymore in one budget after selecting it, there should be some clear big indication with a message telling the user they have already finished the voting and can no longer vote in the selected budget. The reasons may vary, depending on the workflow (so maybe the workflow should control this message?), e.g.
  * The user has already voted in the maximum amount of budgets they are allowed to vote in
  * The user is only able to vote in the budgets for their area e.g. based on their authorization's postal code.
- It would be great if we could set the component permissions settings per budget. Some of the budgets can be limited e.g. only to some area of the city.
- I'd like to be able to set some kind of "metadata" to each budget in order to define e.g.
  * For which postal codes this budget is for
  * The "zone code" for the area this budget is for
  * The name of the area of the city this budget is for
  * etc.
- I'd also like to have some kind of way to display only the budgets on the front page where the user is allowed to vote in, they don't see the reset. This would require implementing the budget specific component permissions.

### UI Improvements

- After selecting one budget, there should be a back link to the budgets listing view and possibly a breadcrumbs indicator indicating where you are
- I'd like the budgets listing section to be full width 100% with the budget names emphasized with a larger font
- I'd like to have some kind of an arrow icon to indicate that you can go "inside" each budget
- I'd like the budget description to show already on the budgets listing page

agustibr commented:
Thanks @ahukkanen for the review 😄.

> ### Bugs

- Fixed the missing translation on component permissions:  now the i18n key should be in plural form (budgets).
- Fixed undefined method `budget`: fixed it by changing `budget.budget` to `budget`.

> ### Feature requests

- > I think it should be possible to programmatically "register" new budget workflows. Currently I need to override the whole method (Decidim::Budgets.workflows)  

  It's possible to add new workflows to the existing, this is documented in the decidim-budgets [README](https://github.com/decidim/decidim/blob/feature/budget_component_w_many_budgets/decidim-budgets/README.md#adding-a-custom-workflow) also as an example the `decidim-generator` adds the `random` workflow to the `development_app`, not present in the `decidim-budgets` module.

- > If the workflow tells voting is not possible anymore in one budget after selecting it, there should be some clear big indication with a message telling the user they have already finished the voting and can no longer vote in the selected budget. The reasons may vary, depending on the workflow (so maybe the workflow should control this message?), e.g.
    > - The user has already voted in the maximum amount of budgets they are allowed to vote in
    > - The user is only able to vote in the budgets for their area e.g. based on their authorization's postal code.  

  Yes, a callout/alert has been added after [@decidim/product 's review](https://github.com/decidim/decidim/issues/5715#issuecomment-670084116), the `cell("decidim/budgets/limit_announcement")` is responsible for this, with 2 possible messages:
  - If the user has to discard to vote:  
    <img width="759" alt="Screenshot 2020-08-24 at 17 23 32" src="https://user-images.githubusercontent.com/210216/91063719-a1767300-e62e-11ea-8929-d7ccfb60945f.png">
  - If the user can't vote on the budget:
    <img width="637" alt="Screenshot 2020-08-24 at 17 22 55" src="https://user-images.githubusercontent.com/210216/91063781-b18e5280-e62e-11ea-8c61-92f582df79d0.png">
- > It would be great if we could set the component permissions settings per budget. Some of the budgets can be limited e.g. only to some area of the city.
- > I'd like to be able to set some kind of "metadata" to each budget in order to define e.g.
  > - For which postal codes this budget is for
  > - The "zone code" for the area this budget is for
  > - The name of the area of the city this budget is for
  > - etc.  
  
  This two previous comments, are probably solved with PR #6309 that adds scopes to the Budgets Component and its resources (feel free to review it too 😄), so a workflow can (dis)allow a user to vote within the budget's scope depending on their authorization's scope.
  Please read this comment on [the flow for permissions and the Budgets component](https://github.com/decidim/decidim/issues/5715#issuecomment-670447309)

- > I'd also like to have some kind of way to display only the budgets on the front page where the user is allowed to vote in, they don't see the reset. This would require implementing the budget specific component permissions.  
  
  Currently you would have to ovewrite (in your app) the budget card_list found in `app/cells/decidim/budgets/budgets_list/card_list.erb` and only loop through the `allowed` budgets collection or alternatively check with the `current_workflow.vote_allowed?(budget)` method.

> ### UI Improvements

- > After selecting one budget, there should be a back link to the budgets listing view and possibly a breadcrumbs indicator indicating where you are.  

  Currently the back link can be found in the more information modal.
  
- > I'd like the budgets listing section to be full width 100% with the budget names emphasized with a larger font
- > I'd like to have some kind of an arrow icon to indicate that you can go "inside" each budget
- > I'd like the budget description to show already on the budgets listing page  
  
  This can be discussed in the [Improve the Budgets component Epic](https://github.com/decidim/decidim/issues/5709).

Cheers 😁

ahukkanen commented:
I found another bug in development environment after upgrading an existing instance:
- The component does not have the `total_budget` global setting defined in the database
- Previously it has been using the default value because this setting is not defined
- In this case, you will get a `NaN` error on the budgets page because the migration assumes this is always defined

If you check this current line:
https://github.com/decidim/decidim/blob/e527b7c18d34e3586f7309d785b234c0e0ccaba3/decidim-budgets/app/models/decidim/budgets/order.rb#L56

I believe it will fetch the default through `component.settings.total_budget` when that is not defined for the budget (or at least it has worked before which is why I assume so).

Now, looking at this migration and particularly this line in it:
https://github.com/decidim/decidim/blob/7b74c845a038ff1ea3990a9a91454bd3cd317f3d/decidim-budgets/db/migrate/20200714103519_move_budgets_to_own_model.rb#L54

This does not have a fallback for the "default" value when this configuration is not defined.

In addition, you could add a validation condition to the budget model to force it to always have the `total_budget` defined. If this was defined, this kind of issue would've been noticed already during the migration.

agustibr commented:
Thanks @ahukkanen for pointing this out 😄.

Now the migration sets the previously default value for the `:total_budget` when it's not defined.

https://github.com/decidim/decidim/blob/3b5627eced22e3d0e5063f4eca208260a0a4a8a5/decidim-budgets/db/migrate/20200714103519_move_budgets_to_own_model.rb#L54-L58

ahukkanen commented:
Thanks for the comments, clarifications and further fixes @agustibr!

I'll add my notes about changing the budgets listing UI with some examples to #5709, thanks also for pointing it out!

Here's few other notes regarding the UI after further "test driving":

### Bug: budgets sorting

The budgets are not listed by their `weight` attribute in the participant view or the admin panel view:

https://github.com/decidim/decidim/blob/320f00479f663ef1e24d1e840ed9aa37f305e387/decidim-budgets/app/controllers/decidim/budgets/admin/budgets_controller.rb#L73

https://github.com/decidim/decidim/blob/320f00479f663ef1e24d1e840ed9aa37f305e387/decidim-budgets/lib/decidim/budgets/workflows/base.rb#L119

### Bug: more information modal is visible when it is not defined

I noticed the "more information" link is visible with the empty modal window even when the text is not defined for that.

I believe you should add a check here:

https://github.com/decidim/decidim/blob/320f00479f663ef1e24d1e840ed9aa37f305e387/decidim-budgets/app/cells/decidim/budgets/budget_information_modal_cell.rb#L6

As follows:

```ruby
  def show
    return if more_information.blank?

    render
  end
```

### About the "back to" links

You mentioned the following regarding the "back to" link:
> Currently the back link can be found in the more information modal.

I have noticed this but it is extremely difficult to find from there. I wanted to suggest adding it on top of the budgeting view as follows:

![decidim-budgets-back-to-all](https://user-images.githubusercontent.com/864340/91642586-387b6a80-ea35-11ea-88bd-f62f772967dd.png)

This should be only visible when there are more than 1 budget.

Another consistent note we get from UI experts is that the wording "back to" or "back" should be avoided in such links because you cannot assume where the user came from. They could come e.g. from a lift in the front page, from the omniauth banner on top of the site or from a search engine. In such a situation, it can be confusing to the user to use the word "back" because they are not going "back" to where they came from.

### Confusing terminology

Another change I'd like to suggest is for this term:
https://github.com/decidim/decidim/blob/320f00479f663ef1e24d1e840ed9aa37f305e387/decidim-budgets/config/locales/en.yml#L108

I think the wording "active votes" can be confusing to non-technical users. I'd suggest something like:

> You have already voted in %{links} or you have started to vote on another budget. To vote on this budget you must <a href="%{landing_path}">delete your vote and start over</a>.

If this does not sound any better, feel free to come up with a better alternative. I'd just like to get rid of the term "active votes" because that be confusing.

agustibr commented:
Hi @ahukkanen, thanks again! 😁

The bug _sorting budgets by weight_ is solved in [commit a5961d1](https://github.com/decidim/decidim/pull/6365/commits/a5961d151b8b4853c71cd39ddf28971882c099ed) in PR #6365.

The suggestions, make sense but I think this should be discussed with @decidim/product as this PR implements the UI approved and present in the `decidim_app-design`, so I encourage you to open an issue to discuss it further to improve usability.

agustibr commented:
@decidim/core this has been approved by @decidim/product, can you review it please? 😄

ahukkanen commented:
@agustibr OK, I understand regarding the other issues, I'll move them to #5709.

But regarding the more information modal visibility, I believe it could be fixed already for this?

agustibr commented:
> But regarding the more information modal visibility, I believe it could be fixed already for this?

@ahukkanen I didn't hide the more information modal, when there's no info, because it also has the _back_ link inside, it would make total sense to hide it if the back link was outside the modal as you proposed.

ahukkanen commented:
Few more bugs I found:

### Error when user enters invalid budget or project IDs to the URL

When the user modifies the URL manually and enters invalid ID numbers for budgets or projects, there will be a "server error" on the page. I believe better would be to return a 404 in these cases, e.g. see:

https://github.com/decidim/decidim/blob/bc7053beaf1b6016be42dcd554ff753cbe1006bb/decidim-proposals/app/controllers/decidim/proposals/proposals_controller.rb#L58

This applies to both `index` and `show` actions in `Decidim::Budgets::ProjectsController`. Also, a similar check should be added to the `show` action in Decidim::Budgets::BudgetsController.

### The user's public timeline is completely broken if the user has left a comment on a budgeting project

- Log in as a user
- Make sure budgeting comments are enabled
- Go to the budgets and open one project
- Leave a comment in that project
- Now, go to your public profile => you will be redirected to the timeline view
- You will see an error on that page (undefined method `project_path`)

I guess you would need to implement the `project_path` method on the budgets engine router. Apparently this does not work either (that you've tried?):
https://github.com/decidim/decidim/commit/f1fafe535a1f075fc4315daaaac352d31917470b#diff-9477c4e74353f50121762aadf11c0951L35

tramuntanal reviewed (approved):
@agustibr this is a very clean and easy to understand refactor even with the many changes it implied, congratulations!

tramuntanal commented:
Thank you @ahukkanen for the heavy testing you're doing on this feature!
@agustibr can you fix the two problems found?
- [ ] Error when user enters invalid budget or project IDs to the URL
- [ ] The user's public timeline is completely broken if the user has left a comment on a budgeting project
  - [ ] This bug also seems to affect the front page timeline block
thanks!

agustibr commented:
Thanks @tramuntanal 😁! (and @ahukkanen)

Yes, I'll push fixes for both.

agustibr commented:
> - [x] Error when user enters invalid budget or project IDs to the URL
> - [x] The user's public timeline is completely broken if the user has left a comment on a budgeting project
>   - [x] This bug also seems to affect the front page timeline block

@tramuntanal the bugs are solved 😁 , the push dismissed your approval review.

agustibr commented:
> I guess you would need to implement the project_path method on the budgets engine router. Apparently this does not work either (that you've tried?):
f1fafe5#diff-9477c4e74353f50121762aadf11c0951L35

@ahukkanen since #6274 was merged `resource_locator` is able to generate the path for nested resources:

```
resource_locator([project.budget, project]).path
# => "/processes/ad-sunt/f/4/budgets/1/projects/1"
```

So I've refactored to use the `resource_locator` with nested resources.
